xarray/core/computation.py

```diff
@@ -0,0 +1,261 @@
+from collections import NamedTuple
+
+
+def result_name(objects):
+    # use the same naming heuristics as pandas:
+    # https://github.com/blaze/blaze/issues/458#issuecomment-51936356
+    names = set(getattr(obj, 'name', None) for obj in objects)
+    names.discard(None)
+    if len(names) == 1:
+        name, = names
+    else:
+        name = None
+    return name
+
+
+def apply_dataarray(func, args, join='inner', kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+
+    args = deep_align(*args, join=join, copy=False, raise_on_invalid=False)
+
+    coord_variables = [getattr(getattr(a, 'coords', {}), 'variables')
+                       for a in args]
+    coords = merge_coords_without_align(coord_variables)
+    name = result_name(args)
+
+    data_vars = [getattr(a, 'variable') for a in args]
+    variable = func(*data_vars, **kwargs)
+
+    return DataArray(variable, coords, name=name, fastpath=True)
+
+
+def join_dict_keys(objects, how='inner')
+    joiner = _get_joiner(how)
+    all_keys = (obj.keys() for obj in objects if hasattr(obj, 'keys'))
+    result_keys = list(joiner(pd.Index(keys) for keys in all_keys))
+
+
+def collect_dict_values(objects, keys, fill_value=None)
+    result_values = []
+    for key in keys:
+        values = []
+        for obj in objects:
+            if hasattr(obj, 'keys'):
+                values.append(obj.get(key, fill_value))
+            else:
+                values = obj
+        result_values.append(values)
+    return result_values
+
+
+def apply_dataset(func, args, join='inner', fill_value=None, kwargs=None,
+                  combine_attrs=None):
+    if kwargs is None:
+        kwargs = {}
+
+    if combine_attrs is None:
+        combine_attrs = lambda func, attrs: None
+
+    attrs = combine_attrs(getattr(func, 'func', func),
+                          [getattr(a, 'attrs') for a in args])
+
+    args = deep_align(*args, join=join, copy=False, raise_on_invalid=False)
+
+    coord_variables = [getattr(getattr(a, 'coords', {}), 'variables')
+                       for a in args]
+    coords = merge_coords_without_align(coord_variables)
+
+    list_of_data_vars = [getattr(a, 'data_vars', {}) for a in args]
+    names = join_dict_keys(list_of_data_vars, how=join)
+
+    list_of_variables = [getattr(a, 'variables', {}) for a in args]
+    lists_of_args = reindex_dict_values(list_of_variables, names, fill_value)
+
+    result_vars = OrderedDict()
+    for name, variable_args in zip(names, lists_of_args):
+        result[name] = func(*variable_args, **kwargs)
+    result_vars.update(coords)
+
+    return Dataset._from_vars_and_coord_names(result_vars, coords, attrs)
+
+
+def _calculate_unified_dim_sizes(variables):
+    dim_sizes = OrderedDict()
+
+    for var in variables:
+        try:
+            var_dims = var.dims
+        except AttributeError:
+            continue
+
+        if len(set(var_dims)) < len(var_dims):
```

Comment:
Nope, this should be fine -- when we catch the `AttributeError`, we `continue` to the next item in the loop.